## Importamos la libreria 
Numpy que es la librería fundamental de Python para el cálculo científico y análisis de datos.

In [ ]:
# Importamos la librería NumPy y utilizamos alias como np.
# Se utiliza para trabajar con arreglos, matrices y operaciones matemáticas.
import numpy as np

## CARGA DEL ARCHIVO CSV

In [23]:

# np.genfromtxt permite leer datos incluso si hay errores o valores vacíos
# El uso de genfromtxt() proporciona una forma segura y flexible de importar información
# ya que permite trabajar incluso cuando el archivo contiene celdas vacías o registros incompletos.
cantidad, precio_unitario = np.genfromtxt(
    'ventas_techventas.csv',

    # Indica que está separado por comas
    delimiter=',',

    # Salta la fila de encabezados
    skip_header=1,

    # Nos permite solo seleccionar columnas necesarias
    # columna 7 = cantidad
    # columna 8 = precio_unitario
    usecols=(7, 8),

    # unpack=True separa columnas en variables
    # cantidad, precio_unitario
    unpack=True,

    # Evita que el programa se caiga si hay filas dañadas
    invalid_raise=False,

    # Codificación del archivo
    encoding='latin1',

    # Si hay celdas vacías o datos inválidos los convierte en NaN 
    filling_values=np.nan
)

print ("Cantidad: " ,cantidad)
print()
print("Precio uniatrio: ", precio_unitario)

Cantidad:  [ 2. 15.  5. 10.  3. 20.  2.  8. 12.  1.  4. 30.  5.  3. 25.  2.  6. 18.
 12.  2.  4. 30.  8.  3.  4. 20. 15.  6.  1. 10.  3. 22.  5. 25.  2.  4.
 12. 18.  2.  3. 15.  5.  2.  3. 20.  6. 18. -1.  4. 25.  3.  2. 20. 35.
  5.  2.  7.  9. 20.  4.]

Precio uniatrio:  [1200.   25.  350.   85. 1200.   95.  350.  120.   95. 1200.  350.   25.
   85.  120.   95. 1200.  350.   95.   25. 1200.  120.   95.   85.  350.
 1200.   25.   95.  120. 1200.   85.  350.   95. 1200.   25.  120.  350.
   85.   95.  350. 1200.   25.  120.  350. 1200.   95.   85.   25. 1200.
  120.   95.  350. 1200.   25.   95.  350. 1200.  120.   85.   95.  350.]


## LIMPIEZA DE DATOS

In [32]:
# Convertir negativos a positivos primero
cantidad = np.abs(cantidad)
precio_unitario = np.abs(precio_unitario)

# Una máscara es un arreglo de valores verdaderos que se utiliza 
# para filtrar o seleccionar elementos específicos de una colección de datos. 
mascara_validos = (

    # Para identificar registros con valores nulos 
    # Se aplicó el operador ~ para conservar únicamente los registros que contienen información válida
    ~np.isnan(cantidad) &
    ~np.isnan(precio_unitario) &

    # Validacion de datos negativos 
    (cantidad >= 0) &
    (precio_unitario >= 0)
)

# Aplicar limpieza
cantidad_limpia = cantidad[mascara_validos]
precio_limpio = precio_unitario[mascara_validos]

print ("Cantidad limpia: " ,cantidad_limpia)
print()
print("Precio uniatrio limpio : ", precio_limpio)


Cantidad limpia:  [ 2. 15.  5. 10.  3. 20.  2.  8. 12.  1.  4. 30.  5.  3. 25.  2.  6. 18.
 12.  2.  4. 30.  8.  3.  4. 20. 15.  6.  1. 10.  3. 22.  5. 25.  2.  4.
 12. 18.  2.  3. 15.  5.  2.  3. 20.  6. 18.  1.  4. 25.  3.  2. 20. 35.
  5.  2.  7.  9. 20.  4.]

Precio uniatrio limpio :  [1200.   25.  350.   85. 1200.   95.  350.  120.   95. 1200.  350.   25.
   85.  120.   95. 1200.  350.   95.   25. 1200.  120.   95.   85.  350.
 1200.   25.   95.  120. 1200.   85.  350.   95. 1200.   25.  120.  350.
   85.   95.  350. 1200.   25.  120.  350. 1200.   95.   85.   25. 1200.
  120.   95.  350. 1200.   25.   95.  350. 1200.  120.   85.   95.  350.]


Ahora nuestros datos estan limpios

# NumPy & Fundamentos Python

### Álgebra Lineal Aplicada
- Crea arrays de cantidades y precios del dataset.
- Calcula el vector de ingresos brutos usando multiplicación matricial con numpy.dot o el operador @.
- Compara con la multiplicación elemento a elemento.

In [33]:
#Calculamos el vector de ingresos brutos usando multiplicación matricial con numpy.dot
ingresos = np.dot(cantidad_limpia, precio_limpio)

print("Ingreso total: ", ingresos)

Ingreso total:  90605.0


In [34]:
#Calculamos el vector de ingresos por venta con multiplicacion de elemento a elemento 
ingreso_elemento = cantidad_limpia * precio_limpio

print("Ingresos por venta: ", ingreso_elemento)

Ingresos por venta:  [2400.  375. 1750.  850. 3600. 1900.  700.  960. 1140. 1200. 1400.  750.
  425.  360. 2375. 2400. 2100. 1710.  300. 2400.  480. 2850.  680. 1050.
 4800.  500. 1425.  720. 1200.  850. 1050. 2090. 6000.  625.  240. 1400.
 1020. 1710.  700. 3600.  375.  600.  700. 3600. 1900.  510.  450. 1200.
  480. 2375. 1050. 2400.  500. 3325. 1750. 2400.  840.  765. 1900. 1400.]


## Comparación de métodos

La multiplicación matricial usando .dot() realiza el producto punto entre ambos vectores, devolviendo un único valor que corresponde al ingreso bruto total de todas las ventas.

Por otro lado, la multiplicación elemento a elemento genera un vector donde cada posición representa el ingreso bruto de una venta individual.

Por tanto:

- Multiplicación matricial → ingreso bruto total del conjunto de ventas. 

- Multiplicación elemento a elemento → ingresos individuales por venta.

## Estadísticas con NumPy
Calcula con NumPy puro sobre el campo precio_unitario:
- Media
- Mediana 
- Desviación estándar 
- Percentiles 25 y 75
Agrega un bloque Markdown explicando qué te dicen estos números sobre el catálogo.

In [35]:
# Media
media = np.mean(precio_limpio)

# Mediana
mediana = np.median(precio_limpio)

# Desviación estándar
desviacion = np.std(precio_limpio)

# Percentiles 25 y 75
percentiles = np.percentile(precio_limpio, [25, 75])

print("Media:", media)
print("Mediana:", mediana)
print("Desviación estándar:", desviacion)
print("Percentil 25:", percentiles[0])
print("Percentil 75:", percentiles[1])

Media: 378.25
Mediana: 120.0
Desviación estándar: 445.0597572236789
Percentil 25: 92.5
Percentil 75: 350.0


## Interpretación de las estadísticas

La media del precio unitario es de 378.25 , lo que indica que en el promedio, los productos del catálogo tienen un costo relativamente alto. 

Sin embargo, la mediana es de 120, mucho menor que la media, lo que sugiere que la mayoría de los productos tienen precios bajos o moderados y que existen algunos productos de precio elevado, que incrementan el promedio general

La desviación estándar de 445.06 es alta, lo que evidencia una gran variabilidad en los precios del catálogo. Esto significa que existen diferencias importantes entre los productos ofrecidos

El percentil 25 (92.5) indica que el 25% de los productos tienen precios iguales o inferiores a este valor, mientras que el percentil 75 (350 ) muestra que el 75% de los productos tienen precios iguales o inferiores a 350. Esto confirma que la mayoría de los productos se concentra en rangos de precio bajos y medios, aunque existen algunos artículos considerablemente más costosos que generan una distribución asimétrica de los precios.


## Reflexión CRISP-DM
En este ejercicio se aplican varias fases del ciclo CRISP-DM.

Primero, en la comprensión del negocio, el objetivo es analizar el comportamiento del catálogo de productos y entender los ingresos generados.

Luego, en la comprensión de los datos, se identifican las variables importantes como cantidades y precios, además de detectar problemas como valores nulos o negativos.

En la preparación de datos, se realiza la limpieza eliminando valores NaN y datos inválidos como números negativos, lo cual es clave para evitar errores en los cálculos.

En la fase de modelado, se aplican operaciones de álgebra lineal con NumPy, como el producto punto para calcular ingresos totales.

En la evaluación, se analizan estadísticas como media, mediana y desviación estándar para entender la distribución de precios.

Finalmente, el despliegue se realiza al presentar los resultados en el notebook de forma clara.

La fase más importante en este caso es la preparación de datos, porque si los datos están sucios o incorrectos, todos los cálculos posteriores serán erróneos y llevarán a conclusiones equivocadas.